src/Kernel/Context.class.st

```diff
@@ -1713,20 +1713,25 @@ Context >> return: value [
 ]
 
 { #category : #controlling }
-Context >> return: value from: aSender [ 
+Context >> return: value from: aSender [
 	"For simulation.  Roll back self to aSender and return value from it.  Execute any unwind blocks on the way.  ASSUMES aSender is a sender of self"
-
 	| newTop context |
 	aSender isDead ifTrue: [
 		^ self send: #cannotReturn: to: self with: {value} super: false ].
 	newTop := aSender sender.
 	context := self findNextUnwindContextUpTo: newTop.
+	(context notNil and: [ context unwindBlock = closureOrNil ]) ifTrue: [
+		"Here the return happens inside the unwind block like in following example: 
+				[ #test ] ensure: [ ^ #returnFromUnwindBlock ]
+		In such cases #findNextUnwindContextUpTo: returns the corresponding #ensure context
+		because it is always a sender of unwind block.
+		Here we detect this case and lookup the next unwind context after the current ensure section (current unwind)"
+		context := context findNextUnwindContextUpTo: newTop ].
```

Comment:
If I understand correctly, before your fix the `aboutToReturn:through:` is sent with the `ensure:` context as context.
Then the fix makes sure that the returned context is the next context to unwind, which is the return in the ensure block.

Did you identify what causes the hanging in the first case?

Reply:
The real bug still exists: when simulation is going into #aboutToReturn:through: the image hangs. 
My fix just removes the one specific case where simulation should not go this way. But the real problem is still here. And it's easy to reproduce:
```Smalltalk
[
   [ 100 + 200 ] ensure: [^600]
] ensure: [ #done ]
```
Step inside [^600] block. Then step over it will hang the image.

Reply:
So the situation is much worser. We can't step over return inside ensure:
```Smalltalk
[ ^100 ] ensure: [ 10+30]
```
Step into block [^100]. Then step over it will hang the image.
I checked also Pharo 7 - same issue

Reply:
So why the image hangs in all those cases? This is very interesting story :) 
The short answer: bugs in the debugger lead to the state of target process where it never ends.

The last call of any process is the termination which ends with #suspend.
```Smalltalk
newProcess
	^Process
		forContext: 
			[self value.
			Processor terminateActive] asContext
		priority: Processor activePriority
```
The block here is actually never returned and it must not return. If you will remove the #terminateActive call then processes will end with BlockCannotReturn error (see cannotReturn:) because VM will not find the sender to continue the process. 

And it is what really happens when we got hanging in the debugger: 
- #terminateActive call does not suspend the current process and simply returns.
- VM calls #cannotReturn: which triggers the debugger:
  - new debugger is requested
  - debugger tries to #suspend activeProcess but it does nothing (see below) and process continues. 
  - as process continues the method #cannotReturn: returns.
- cannotReturn: has no sender and therefore VM calls cannotReturn: again - Infinite recursions began.
  
Now why #terminateActive and #suspend can be not working.

The problem is how debugger simulates the #activeProcess calls during stepping operations:. 
```Smalltalk
process := Process activeProcess
```
The StepOver this expression is evaluated by the debugger process (UIProcess) but still the result will be the process under the debugger (not a UIProcess). 
To achieve this the debugger wraps all stepping operations using #effectiveProcess trick:
```Smalltalk
step
	^Processor activeProcess
		evaluate: [suspendedContext := suspendedContext step]
		onBehalfOf: self
```
The active process here is UIProcess (which handles StepOver click). It sets the variable #effectiveProcess to the process under the debugger. During the block execution the Processor will always use #effectiveProcess as an active one instead of the real active process which executes the debugger command. 
For example you can step over "Processor terminateActive" and it will not terminate the UIProcess while the command is really executed by UIProcess.

Now it should be clear how the normal process completion becomes broken: if #effectiveProcess variable is not reset by the debugger the #terminateActive call will never kill the current process.
And same issue happens with #suspend call. When debugger is requested it uses "Processor activeProcess" to get a process for debugging and to suspend it but it will never return the actual current process. Therefore the #suspend from the debugger will not stop the current process (it will suspend some other process) and the current process will continue execution which will cause the buggy recursion.

Now why effectiveProcess can be not reset? Method #evaluate:onBehalfOf: contains the ensure logic to reset the variable back. 
The block here is supposed to do the context/stack manipulations. And some of them are really not trivial. The processing of unwinds (ensure parts) are their responsibility. So it's obvious that there can be some bugs or missing parts. Especially that we don't have good tests for simulation logic.

For this concrete issue with StepOver the problem is in the way how debugger intercepts the return from the callee. The current trap is not enough for the returns inside unwind contexts. And therefore the simulated code is not stopped on the current context and completely finishes the execution up to last #terminateActive in the #newProcess. So #terminateActive is executed with effective process and it hangs the image.

The trap needs to be improved and maybe it's not really easy to do. But what is more important is to find a general fix for described problem. 
I hope you understand my explanation. I will push some obvious changes soon: #newProcess should ensure the termination of actual active process despite on any assigned effectiveProcess.